Code Review

SredzinskiPatryk opened the pull request.

Commit:
Code Review

vandermesis committed:
Merge remote-tracking branch 'refs/remotes/origin/develop'

Conflicts:
	ColorPatterns/Scenes/Game/GameCreator.swift
	ColorPatterns/Scenes/Game/GameTimer.swift
	ColorPatterns/Scenes/StartMenu/StartMenuController.swift
	ColorPatternsTests/ColorPatternsTests.swift

vandermesis committed:
Code review fixes

Fixed:
	Removed unneeded ColorPatternTests
	Rename K file to Constants
	Change naming convention of cutom colors
	Change naming convention of some properites
	Setup of gesture recognizer in xib
	Move default game time to constants
	Refactor Score methods and properties
	Make userScore read only
	Add TODOs for unfixed issues

vandermesis commented on ColorPatterns/Extensions/CGFloat+Converters.swift:
```diff
@@ -10,6 +10,7 @@ import UIKit
 
 extension CGFloat {
 
+    // TODO: Replace with custom colors? How will I add 1000 custom colors?
```
<table data-meta="generated_by_danger">
  <tbody>
    <tr>
      <td>:warning:</td>
      <td width="100%" data-sticky="false"><span data-href="https://github.com/vandermesis/ColorPatterns/blob/c07cd9a3b7a63e61e86332aeb4fa4cabf8b81c17/ColorPatterns/Extensions/CGFloat+Converters.swift#L13"></span>TODOs should be resolved (Replace with custom colors? Ho…).<br />
<code>todo</code> <code>CGFloat+Converters.swift:13</code></td>
    </tr>
  </tbody>
</table>

vandermesis commented on ColorPatterns/Scenes/Game/GameController.swift:
```diff
@@ -42,23 +42,27 @@ final class GameController: UIViewController {
     }
 
     @IBAction private func colorBarPressed(_ sender: UIButton) {
+        // TODO: Try not to use tags, make an another way to identify buttons
```
<table data-meta="generated_by_danger">
  <tbody>
    <tr>
      <td>:warning:</td>
      <td width="100%" data-sticky="false"><span data-href="https://github.com/vandermesis/ColorPatterns/blob/c07cd9a3b7a63e61e86332aeb4fa4cabf8b81c17/ColorPatterns/Scenes/Game/GameController.swift#L51"></span>TODOs should be resolved (Try not to use tags, make an a…).<br />
<code>todo</code> <code>GameController.swift:51</code></td>
    </tr>
  </tbody>
</table>

vandermesis commented on ColorPatterns/Scenes/Game/GameController.swift:
```diff
@@ -42,23 +42,27 @@ final class GameController: UIViewController {
 
     private func setupGameTimer() {
-        timeLabel.text = gameTimer.timeLeft.formatToString
+        // TODO: You can move timer to viewmodel
```
<table data-meta="generated_by_danger">
  <tbody>
    <tr>
      <td>:warning:</td>
      <td width="100%" data-sticky="false"><span data-href="https://github.com/vandermesis/ColorPatterns/blob/c07cd9a3b7a63e61e86332aeb4fa4cabf8b81c17/ColorPatterns/Scenes/Game/GameController.swift#L64"></span>TODOs should be resolved (You can move timer to viewmode…).<br />
<code>todo</code> <code>GameController.swift:64</code></td>
    </tr>
  </tbody>
</table>

vandermesis commented on ColorPatterns/Scenes/Game/GameController.swift:
```diff
@@ -81,18 +85,9 @@ private extension GameController {
-
     private func playSound(soundFile: Int) {
-        if defaults.bool(forKey: K.DefaultsKeys.sound) == true {
+        // TODO: Make a helper class that does all that stuff
```
<table data-meta="generated_by_danger">
  <tbody>
    <tr>
      <td>:warning:</td>
      <td width="100%" data-sticky="false"><span data-href="https://github.com/vandermesis/ColorPatterns/blob/c07cd9a3b7a63e61e86332aeb4fa4cabf8b81c17/ColorPatterns/Scenes/Game/GameController.swift#L89"></span>TODOs should be resolved (Make a helper class that does …).<br />
<code>todo</code> <code>GameController.swift:89</code></td>
    </tr>
  </tbody>
</table>

vandermesis commented on ColorPatterns/Scenes/Game/GameCreator.swift:
```diff
@@ -12,6 +12,7 @@ struct GameCreator {
 
     func getController(difficulty: Level) -> GameController {
 
+        // TODO: How about moving timer to viewModel? seems like it is not related to UI
```
<table data-meta="generated_by_danger">
  <tbody>
    <tr>
      <td>:warning:</td>
      <td width="100%" data-sticky="false"><span data-href="https://github.com/vandermesis/ColorPatterns/blob/c07cd9a3b7a63e61e86332aeb4fa4cabf8b81c17/ColorPatterns/Scenes/Game/GameCreator.swift#L15"></span>TODOs should be resolved (How about moving timer to view…).<br />
<code>todo</code> <code>GameCreator.swift:15</code></td>
    </tr>
  </tbody>
</table>

vandermesis commented on ColorPatterns/Scenes/Game/GameSounds.swift:
```diff
@@ -12,13 +12,15 @@ struct GameSounds {
+                                             withExtension: Constants.Sounds.wavFormat) else { return }
         do {
             try audioPlayer = AVAudioPlayer(contentsOf: soundURL)
+            // TODO: You're creating new audio player for each sound? why?
```
<table data-meta="generated_by_danger">
  <tbody>
    <tr>
      <td>:warning:</td>
      <td width="100%" data-sticky="false"><span data-href="https://github.com/vandermesis/ColorPatterns/blob/c07cd9a3b7a63e61e86332aeb4fa4cabf8b81c17/ColorPatterns/Scenes/Game/GameSounds.swift#L22"></span>TODOs should be resolved (You’re creating new audio play…).<br />
<code>todo</code> <code>GameSounds.swift:22</code></td>
    </tr>
  </tbody>
</table>

vandermesis commented on ColorPatterns/Scenes/Game/GameViewModel.swift:
```diff
@@ -14,10 +14,11 @@ protocol GameViewModelDelegate: class {
 
 final class GameViewModel {
 
+    // TODO: Make a protocol for this class and then this class implementing that protocol, so you can use it in controller
```
<table data-meta="generated_by_danger">
  <tbody>
    <tr>
      <td>:warning:</td>
      <td width="100%" data-sticky="false"><span data-href="https://github.com/vandermesis/ColorPatterns/blob/c07cd9a3b7a63e61e86332aeb4fa4cabf8b81c17/ColorPatterns/Scenes/Game/GameViewModel.swift#L17"></span>TODOs should be resolved (Make a protocol for this class…).<br />
<code>todo</code> <code>GameViewModel.swift:17</code></td>
    </tr>
  </tbody>
</table>

vandermesis commented on ColorPatterns/Scenes/Game/GameViewModel.swift:
```diff
@@ -54,7 +55,7 @@ private extension GameViewModel {
     }
 
-    //TODO: Remove prints when not needed
+    // TODO: Remove prints when not needed
```
<table data-meta="generated_by_danger">
  <tbody>
    <tr>
      <td>:warning:</td>
      <td width="100%" data-sticky="false"><span data-href="https://github.com/vandermesis/ColorPatterns/blob/c07cd9a3b7a63e61e86332aeb4fa4cabf8b81c17/ColorPatterns/Scenes/Game/GameViewModel.swift#L58"></span>TODOs should be resolved (Remove prints when not needed).<br />
<code>todo</code> <code>GameViewModel.swift:58</code></td>
    </tr>
  </tbody>
</table>

vandermesis commented on ColorPatterns/Scenes/GameOver/GameOverController.swift:
```diff
@@ -29,13 +29,14 @@ final class GameOverController: UIViewController {
 
     @IBAction private func restartGameButtonPressed(_ sender: UIButton) {
-        self.view.window?.rootViewController?.dismiss(animated: true, completion: nil)       
+        //TODO: whoooho, make sure you know what you're doing, plan your navigation stack always and be always sure what is where, you'd need to change it
```
<table data-meta="generated_by_danger">
  <tbody>
    <tr>
      <td>:warning:</td>
      <td width="100%" data-sticky="false"><span data-href="https://github.com/vandermesis/ColorPatterns/blob/c07cd9a3b7a63e61e86332aeb4fa4cabf8b81c17/ColorPatterns/Scenes/GameOver/GameOverController.swift#L32"></span>TODOs should be resolved (whoooho, make sure you know wh…).<br />
<code>todo</code> <code>GameOverController.swift:32</code></td>
    </tr>
  </tbody>
</table>

vandermesis commented on ColorPatterns/Scenes/StartMenu/StartMenuController.swift:
```diff
@@ -24,6 +24,8 @@ final class StartMenuController: UIViewController {
     }
     
     @IBAction private func difficultyButtonPressed(_ sender: UIButton) {
+        // TODO: it's not a good idea to use tags to do such things I'd go with enum difficulty: easy, med, hard and setup buttons with these values, but not use tags
```
<table data-meta="generated_by_danger">
  <tbody>
    <tr>
      <td>:warning:</td>
      <td width="100%" data-sticky="false"><span data-href="https://github.com/vandermesis/ColorPatterns/blob/c07cd9a3b7a63e61e86332aeb4fa4cabf8b81c17/ColorPatterns/Scenes/StartMenu/StartMenuController.swift#L27"></span>Line should be 160 characters or less: currently 166 characters<br />
<code>line_length</code> <code>StartMenuController.swift:27</code></td>
    </tr>
  </tbody>
</table>

vandermesis replied:
<table data-meta="generated_by_danger">
  <tbody>
    <tr>
      <td>:warning:</td>
      <td width="100%" data-sticky="false"><span data-href="https://github.com/vandermesis/ColorPatterns/blob/c07cd9a3b7a63e61e86332aeb4fa4cabf8b81c17/ColorPatterns/Scenes/StartMenu/StartMenuController.swift#L27"></span>TODOs should be resolved (it’s not a good idea to use ta…).<br />
<code>todo</code> <code>StartMenuController.swift:27</code></td>
    </tr>
  </tbody>
</table>

vandermesis commented on ColorPatterns/Scenes/StartMenu/StartMenuController.swift:
```diff
@@ -38,13 +40,14 @@ final class StartMenuController: UIViewController {
 private extension StartMenuController {
 
     private func setupDifficultyButtonColor(button: Int) {
+        // TODO: crash if I do:  setupDifficultyButtonColor(button: 52) - this should be forbidden
```
<table data-meta="generated_by_danger">
  <tbody>
    <tr>
      <td>:warning:</td>
      <td width="100%" data-sticky="false"><span data-href="https://github.com/vandermesis/ColorPatterns/blob/c07cd9a3b7a63e61e86332aeb4fa4cabf8b81c17/ColorPatterns/Scenes/StartMenu/StartMenuController.swift#L43"></span>TODOs should be resolved (crash if I do:  setupDifficult…).<br />
<code>todo</code> <code>StartMenuController.swift:43</code></td>
    </tr>
  </tbody>
</table>

vandermesis commented on ColorPatterns/Support/AppDelegate.swift:
```diff
@@ -14,7 +14,9 @@ final class AppDelegate: UIResponder, UIApplicationDelegate {
     func application(_ application: UIApplication, didFinishLaunchingWithOptions launchOptions: [UIApplication.LaunchOptionsKey: Any]?) -> Bool {
-        //  Show lauch screen for 0.5 sec to display logo
+
+        // TODO: Never seen that kind of logic here, but It could be smarter to add subview of launchscreen to your game controller
```
<table data-meta="generated_by_danger">
  <tbody>
    <tr>
      <td>:warning:</td>
      <td width="100%" data-sticky="false"><span data-href="https://github.com/vandermesis/ColorPatterns/blob/c07cd9a3b7a63e61e86332aeb4fa4cabf8b81c17/ColorPatterns/Support/AppDelegate.swift#L18"></span>TODOs should be resolved (Never seen that kind of logic …).<br />
<code>todo</code> <code>AppDelegate.swift:18</code></td>
    </tr>
  </tbody>
</table>

vandermesis commented:
<!--
  0 Errors
  1 Warning: Please provide a summary in th...
  0 Messages
  0 Markdowns
-->
<table>
  <thead>
    <tr>
      <th width="50"></th>
      <th width="100%" data-danger-table="true" data-kind="Warning">
          1 Warning
      </th>
     </tr>
  </thead>
  <tbody>
    <tr>
      <td>:warning:</td>
      <td data-sticky="false">Please provide a summary in the Pull Request description</td>
    </tr>
  </tbody>
</table>

<p align="right" data-meta="generated_by_danger">
  Generated by :no_entry_sign: <a href="https://danger.systems/">Danger</a>
</p>

vandermesis committed:
Merge remote-tracking branch 'refs/remotes/origin/develop'

Conflicts:
	ColorPatterns/Scenes/GameOver/GameOverController.xib

vandermesis committed:
Select default difficulty button in StartMenu xib

vandermesis committed:
Fix GameTimer unit tests after refactoring

vandermesis commented on ColorPatterns/Scenes/StartMenu/StartMenuController.swift:
```diff
@@ -24,6 +24,9 @@ final class StartMenuController: UIViewController {
     }
     
     @IBAction private func difficultyButtonPressed(_ sender: UIButton) {
+        // TODO: it's not a good idea to use tags to do such things
```
<table data-meta="generated_by_danger">
  <tbody>
    <tr>
      <td>:warning:</td>
      <td width="100%" data-sticky="false"><span data-href="https://github.com/vandermesis/ColorPatterns/blob/6055d8cf2ff9610160e4bc6d457c28b6db6bd01c/ColorPatterns/Scenes/StartMenu/StartMenuController.swift#L27"></span>TODOs should be resolved (it’s not a good idea to use ta…).<br />
<code>todo</code> <code>StartMenuController.swift:27</code></td>
    </tr>
  </tbody>
</table>

vandermesis commented on ColorPatterns/Scenes/StartMenu/StartMenuController.swift:
```diff
@@ -38,13 +41,14 @@ final class StartMenuController: UIViewController {
 private extension StartMenuController {
 
     private func setupDifficultyButtonColor(button: Int) {
+        // TODO: crash if I do:  setupDifficultyButtonColor(button: 52) - this should be forbidden
```
<table data-meta="generated_by_danger">
  <tbody>
    <tr>
      <td>:warning:</td>
      <td width="100%" data-sticky="false"><span data-href="https://github.com/vandermesis/ColorPatterns/blob/6055d8cf2ff9610160e4bc6d457c28b6db6bd01c/ColorPatterns/Scenes/StartMenu/StartMenuController.swift#L44"></span>TODOs should be resolved (crash if I do:  setupDifficult…).<br />
<code>todo</code> <code>StartMenuController.swift:44</code></td>
    </tr>
  </tbody>
</table>

vandermesis committed:
Move GameTimer to GameViewModel

Changes:
	Add protocols to controller and viewmodel to use DI
	Remove GameViewModel Delegate design pattern

vandermesis committed:
Fix GameViewModel unit tests

Changes:
	Add GameTimer protocol for testability
	Add GameViewModel unit tests for timer

vandermesis commented on ColorPatterns/Scenes/Game/GameController.swift:
```diff
@@ -81,40 +84,10 @@ private extension GameController {
-
     private func playSound(soundFile: Int) {
-        if defaults.bool(forKey: K.DefaultsKeys.sound) == true {
+        // TODO: Make a helper class that does all that stuff
```
<table data-meta="generated_by_danger">
  <tbody>
    <tr>
      <td>:warning:</td>
      <td width="100%" data-sticky="false"><span data-href="https://github.com/vandermesis/ColorPatterns/blob/67f2732f59af0d268971dc68cf5f930b621eb1ad/ColorPatterns/Scenes/Game/GameController.swift#L88"></span>TODOs should be resolved (Make a helper class that does …).<br />
<code>todo</code> <code>GameController.swift:88</code></td>
    </tr>
  </tbody>
</table>

vandermesis commented on ColorPatterns/Scenes/Game/GameViewModel.swift:
```diff
@@ -43,18 +52,18 @@ extension GameViewModel {
     }
 
-    //TODO: Remove prints when not needed
+    // TODO: Remove prints when not needed
```
<table data-meta="generated_by_danger">
  <tbody>
    <tr>
      <td>:warning:</td>
      <td width="100%" data-sticky="false"><span data-href="https://github.com/vandermesis/ColorPatterns/blob/67f2732f59af0d268971dc68cf5f930b621eb1ad/ColorPatterns/Scenes/Game/GameViewModel.swift#L66"></span>TODOs should be resolved (Remove prints when not needed).<br />
<code>todo</code> <code>GameViewModel.swift:66</code></td>
    </tr>
  </tbody>
</table>

vandermesis committed:
Refactor GameSounds

Changes:
	Move GameSounds to GameViewModel
	Move UserDefaults to GameSounds
	Move sound settings check to GameSounds
	Refactor AudioPlayer init to lazy init
	Init UserDefaults in GameCreator

vandermesis committed:
Fix GameViewModel tests after GameSound refactoring

vandermesis commented on ColorPatterns/Scenes/Game/GameController.swift:
```diff
@@ -42,33 +38,34 @@ final class GameController: UIViewController {
     }
 
     @IBAction private func colorBarPressed(_ sender: UIButton) {
+        // TODO: Try not to use tags, make an another way to identify buttons
```
<table data-meta="generated_by_danger">
  <tbody>
    <tr>
      <td>:warning:</td>
      <td width="100%" data-sticky="false"><span data-href="https://github.com/vandermesis/ColorPatterns/blob/c279b1ccceb81114848aa1e590426203038db92d/ColorPatterns/Scenes/Game/GameController.swift#L45"></span>TODOs should be resolved (Try not to use tags, make an a…).<br />
<code>todo</code> <code>GameController.swift:45</code></td>
    </tr>
  </tbody>
</table>

vandermesis commented on ColorPatterns/Scenes/Game/GameSounds.swift:
```diff
@@ -8,20 +8,52 @@
+                                       withExtension: Constants.Sounds.wavFormat)
+        return soundURL
+    }
+    //FIXME: That doesn't work properly.
```
<table data-meta="generated_by_danger">
  <tbody>
    <tr>
      <td>:warning:</td>
      <td width="100%" data-sticky="false"><span data-href="https://github.com/vandermesis/ColorPatterns/blob/c279b1ccceb81114848aa1e590426203038db92d/ColorPatterns/Scenes/Game/GameSounds.swift#L19"></span>FIXMEs should be resolved (That doesn’t work properly.).<br />
<code>todo</code> <code>GameSounds.swift:19</code></td>
    </tr>
  </tbody>
</table>

vandermesis commented on ColorPatterns/Scenes/Game/GameSounds.swift:
```diff
@@ -8,20 +8,52 @@
+    //FIXME: That doesn't work properly.
+    // Only some sounds are played because only one player is created and it has to play sound to the end
+    private lazy var audioPlayer: AVAudioPlayer = {
+        [unowned self] in
```
<table data-meta="generated_by_danger">
  <tbody>
    <tr>
      <td>:warning:</td>
      <td width="100%" data-sticky="false"><span data-href="https://github.com/vandermesis/ColorPatterns/blob/c279b1ccceb81114848aa1e590426203038db92d/ColorPatterns/Scenes/Game/GameSounds.swift#L22"></span>It’s safer to use weak instead of unowned</td>
    </tr>
  </tbody>
</table>

vandermesis commented on ColorPatterns/Scenes/Game/GameViewModel.swift:
```diff
@@ -36,25 +48,26 @@ extension GameViewModel {
     }
 
-    //TODO: Remove prints when not needed
+    // TODO: Remove prints when not needed
```
<table data-meta="generated_by_danger">
  <tbody>
    <tr>
      <td>:warning:</td>
      <td width="100%" data-sticky="false"><span data-href="https://github.com/vandermesis/ColorPatterns/blob/f4a732c9facc85e6fc76fb056b49a92b6ae75a06/ColorPatterns/Scenes/Game/GameViewModel.swift#L70"></span>TODOs should be resolved (Remove prints when not needed).<br />
<code>todo</code> <code>GameViewModel.swift:70</code></td>
    </tr>
  </tbody>
</table>

vandermesis committed:
Change StartMenu buttons access methods

vandermesis committed:
Fix StartMenuViewModel unit tests

vandermesis committed:
Change buttons access method in GameController

vandermesis committed:
Refactor showing logo on launch method with custom view

vandermesis committed:
Revert GameSound logic back to previous solution

vandermesis committed:
Change GameOver scene navigation to StartMenu method

vandermesis committed:
Fix GameOverCreator unit tests

vandermesis committed:
Remove unneeded TODOs

vandermesis committed:
Handle hardcoded string values

vandermesis committed:
Refactor showing logo on launch method with LaunchScreen
